src/heap.h

```diff
@@ -5,28 +5,114 @@ template<class ValueType>
 class Heap
 {
 private:
-    int heapSize;
     Node<ValueType>* root;
+    int size;
+    int height;
 
-protected:
+    int calculateHeapHeight(int size);
+    Node<ValueType>* insertValue(ValueType value, Node<ValueType>* node, int level);
+    void rebuild(Node<ValueType>* node);
     void swapNodes(Node<ValueType>* node1, Node<ValueType>* node2);
+    void deleteHeap(Node<ValueType>* node);
+
+protected:
+    bool virtual compare(Node<ValueType>* node, Node<ValueType>* toCompare) = 0;
 
 public:
     Heap();
-    virtual ~Heap() = default;
+    virtual ~Heap();
 
     void show();
-    void virtual insertValue(ValueType value) = 0;
+    void insertValue(ValueType value);
     void virtual build(ValueType values[]) = 0;
     ValueType virtual extract() = 0;
     ValueType virtual get() = 0;
 };
 
 template<class ValueType>
-Heap<ValueType>::Heap() : heapSize(0), root(nullptr) {}
+Heap<ValueType>::Heap() : size(0), height(0), root(nullptr) {}
+
+template<class ValueType>
+void Heap<ValueType>::insertValue(ValueType value) {
+    size++;
+    height = calculateHeapHeight(this->size);
+
+    if (root == nullptr) {
+        root = new Node<ValueType>(value);
+    }
+    else {
+        Node<ValueType>* insertedNode = insertValue(value, root, 1);
+        rebuild(insertedNode);
+    }
+}
+
+template<class ValueType>
+int Heap<ValueType>::calculateHeapHeight(int size) {
+    return ceil(log2(size + 1));
+}
 
 template<class ValueType>
-void Heap<ValueType>::swapNodes(Node<ValueType>* node1, Node<ValueType>* node2) {}
+Node<ValueType>* Heap<ValueType>::insertValue(ValueType value, Node<ValueType>* node, int level){
+    if (level == height - 1 && (node->left == nullptr || node->right == nullptr)) {
+        return new Node<ValueType>(value, node);
+    }
+    else if (level == height - 1) {
+        return nullptr;
+    }
+
+    Node<ValueType>* insertedNode = insertValue(value, node->left, level + 1);
+
+    if (insertedNode == nullptr) {
+        insertedNode = insertValue(value, node->right, level + 1);
+    }
+
+    return insertedNode;
+}
+
+template<class ValueType>
+void Heap<ValueType>::rebuild(Node<ValueType>* node) {
+    Node<ValueType>* nodeTemp = node;
+
+    while (nodeTemp->parent != nullptr) {
+        if (compare(nodeTemp, nodeTemp->parent)) {
+            swapNodes(nodeTemp, nodeTemp->parent);
+        }
+        nodeTemp = nodeTemp->parent;
+    }
+}
+
+template<class ValueType>
+void Heap<ValueType>::swapNodes(Node<ValueType>* node1, Node<ValueType>* node2) {
+    ValueType temp = node1->getValue();
+    node1->setValue(node2->getValue());
+    node2->setValue(temp);
+}
+
+template<class ValueType>
+Heap<ValueType>::~Heap() {
+    if (root != nullptr) {
+        deleteHeap(root);
+    }
+}
+
+template<class ValueType>
+void Heap<ValueType>::deleteHeap(Node<ValueType>* node) {
+    if (node->left == nullptr) {
+        node->parent = nullptr;
+        delete node;
+        return;
+    }
+
+    deleteHeap(node->left);
```

Comment:
Tu mi coś nie pasuje. Nie wywali się jak left nie będzie istniał?

Reply:
Co konkretnie masz na myśli?

Reply:
W sumie wyżej masz sprawdzanie czy left istnieje. node->left == nullptr

Reply:
Ale w sumie na samym początku można by sprawdzić czy node != nullptr bo jeżeli tak to w if (node->left == nullptr) wywali błąd.

Reply:
Przed wywoałaniem metody deleteHeap jest sprawdzane czy root jest nullem. Na początku deleteHeap jest warunek zatrzymania rekurencji w postaci sprawdzenai czy left jest nullem, przed usunięciem prawej gałęzi jest sprawdzenie czy prawe dziecko jest nullem.  Czy gdzieś może polecieć jeszcze nullpointerexception?

Reply:
Tylko w niezbyt prawdopodobnym, ale możliwym wypadku jakby ktoś gdzieś wywołał deleteHeap(root) i root byłby nullptr, zamiast usunąć poprzez destruktor - ja bym po prostu przeniosła to sprawdzenie z destruktora tutaj.

Reply:
Jak coś delete jest private, więc może być tylko wywołane z wnętrza klasy Heap. Mogę dodać sprawdzanie czy node nie jest czasem null. Tylko wtedy będzimey mieć po prostu dodatkowe null checkera który w sumie będzie potrebny tylko w momencie kiedy podamy roota